crates/evm/src/execute.rs

```diff
@@ -163,12 +164,239 @@ pub trait BlockExecutorProvider: Send + Sync + Clone + Unpin + 'static {
+    fn execute_transactions(
+        &mut self,
+        block: &BlockWithSenders,
+    ) -> Result<(Vec<Receipt>, u64), Self::Error>;
+
+    /// Applies any necessary changes after executing the block's transactions.
+    fn apply_post_execution_changes(&mut self) -> Result<Vec<Request>, Self::Error>;
+
+    /// Returns a reference to the current state.
+    fn state_ref(&self) -> &State<DB>;
+
+    /// Sets a hook to be called after each state change during execution.
+    fn with_state_hook(&mut self, hook: Option<Box<dyn OnStateHook>>);
+
+    /// Returns the final bundle state.
+    fn finish(&self) -> BundleState;
+}
+
+/// A strategy factory that can create block execution strategies.
+pub trait BlockExecutionStrategyFactory: Send + Sync + Clone + Unpin + 'static {
+    /// Associated strategy type.
+    type Strategy<DB: Database<Error: Into<ProviderError> + Display>>: BlockExecutionStrategy<
+        DB,
+        Error = BlockExecutionError,
+    >;
+
+    /// Creates a strategy using the give database.
+    fn create_strategy<DB>(&self, db: DB) -> Self::Strategy<DB>
+    where
+        DB: Database<Error: Into<ProviderError> + Display>;
+}
+
+impl<F> Clone for GenericBlockExecutorProvider<F>
+where
+    F: Clone,
+{
+    fn clone(&self) -> Self {
+        Self { strategy_factory: self.strategy_factory.clone() }
+    }
+}
+
+/// A generic block executor provider that can create executors using a strategy factory.
+#[allow(missing_debug_implementations)]
+pub struct GenericBlockExecutorProvider<F> {
+    strategy_factory: F,
+}
+
+impl<F> GenericBlockExecutorProvider<F> {
+    /// Creates a new `GenericBlockExecutorProvider` with the given strategy factory.
+    pub const fn new(strategy_factory: F) -> Self {
+        Self { strategy_factory }
+    }
+}
+
+impl<F> BlockExecutorProvider for GenericBlockExecutorProvider<F>
+where
+    F: BlockExecutionStrategyFactory,
+{
+    type Executor<DB: Database<Error: Into<ProviderError> + Display>> =
+        GenericBlockExecutor<F::Strategy<DB>, DB>;
+
+    type BatchExecutor<DB: Database<Error: Into<ProviderError> + Display>> =
+        GenericBatchExecutor<F::Strategy<DB>, DB>;
+
+    fn executor<DB>(&self, db: DB) -> Self::Executor<DB>
+    where
+        DB: Database<Error: Into<ProviderError> + Display>,
+    {
+        let strategy = self.strategy_factory.create_strategy(db);
+        GenericBlockExecutor::new(strategy)
+    }
+
+    fn batch_executor<DB>(&self, db: DB) -> Self::BatchExecutor<DB>
+    where
+        DB: Database<Error: Into<ProviderError> + Display>,
+    {
+        let strategy = self.strategy_factory.create_strategy(db);
+        GenericBatchExecutor::new(strategy)
+    }
+}
+
+/// A generic block executor that uses a [`BlockExecutionStrategy`] to
+/// execute blocks.
+#[allow(missing_debug_implementations, dead_code)]
+pub struct GenericBlockExecutor<S, DB>
+where
+    S: BlockExecutionStrategy<DB>,
+{
+    strategy: RefCell<S>,
```

Comment:
why do we need a refcell here?

Reply:
> why do we need a refcell here?

sure, this RefCell allows us to have a mut strategy inside the block execution methods (like here https://github.com/paradigmxyz/reth/pull/11447/files#diff-eb5f2c91d277e8bdf71850182110f34c23bcb1d70a675f570572f18af090a1d1R290) without having a mut self receiver here https://github.com/paradigmxyz/reth/pull/11447/files#diff-eb5f2c91d277e8bdf71850182110f34c23bcb1d70a675f570572f18af090a1d1R287

The reason for not creating the strategy in the execute methods is bc we don't have direct access to the state or db at that point (unless we have stored it). AIUI would force us to Clone it or do some std::mem::take or Option::take. 

I started with this latter approach but soon it got overcomplicated compared to what we have now, I can give it another try, let me know wdyt

Reply:
hmm, but this function takes ownership of `self`, so we can consume it entirely via `mut self` and don't need the refcell?

Reply:
absolutely, with `mut self` in `Executor` methods we can remove the `RefCell`, done here https://github.com/paradigmxyz/reth/pull/11447/commits/bce269c233ca33a7c8c1d18cecd1c34d67570385